ddtrace/bootstrap/sitecustomize.py

```diff
@@ -2,30 +2,112 @@
+    elif sys.version_info >= (3, 3):
+        return importlib.find_loader(module_name)
+    elif sys.version_info >= (3, 1):
+        return importlib.find_module(module_name)
+    elif sys.version_info >= (2, 7):
+        try:
+            imp.find_module(module_name)
+        except ImportError:
+            return False
+        else:
+            return True
+    return False
+
+
+def should_cleanup_loaded_modules():
+    dd_unload_sitecustomize_modules = os.getenv("DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE", default="0").lower()
+    if dd_unload_sitecustomize_modules == "0":
+        return False
+    elif dd_unload_sitecustomize_modules not in ("1", "auto"):
+        return False
+    elif dd_unload_sitecustomize_modules == "auto" and not any(
+        is_installed(module_name) for module_name in MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED
+    ):
+        return False
+    return True
+
+
+def cleanup_loaded_modules(aggressive=False):
+    """
+    "Aggressive" here means "cleanup absolutely every module that has been loaded since startup".
+    Non-aggressive cleanup entails leaving untouched certain modules
+    This distinction is necessary because this function is used both to prepare for gevent monkeypatching
+    (requiring aggressive cleanup) and to implement "module cloning" (requiring non-aggressive cleanup)
+    """
+    modules_loaded_since_startup = set(_ for _ in sys.modules if _ not in MODULES_LOADED_AT_STARTUP)
+    # Figuring out modules_loaded_since_startup is necessary because sys.modules has more in it than just what's in
+    # import statements in this file, and unloading some of them can break the interpreter.
+    modules_to_cleanup = modules_loaded_since_startup - MODULES_TO_NOT_CLEANUP
+    if aggressive:
+        modules_to_cleanup = modules_loaded_since_startup
+    # Unload all the modules that we have imported, except for ddtrace and a few
+    # others that don't like being cloned.
+    # Doing so will allow ddtrace to continue using its local references to modules unpatched by
+    # gevent, while avoiding conflicts with user-application code potentially running
+    # `gevent.monkey.patch_all()` and thus gevent-patched versions of the same modules.
+    for m in modules_to_cleanup:
+        if not aggressive and any(
+            m.startswith("%s." % module_to_not_cleanup) for module_to_not_cleanup in MODULES_TO_NOT_CLEANUP
+        ):
+            continue
+        del sys.modules[m]
+
+    if "time" in sys.modules:
+        del sys.modules["time"]
+
+
+will_run_module_cloning = should_cleanup_loaded_modules()
+if not will_run_module_cloning:
+    # Perform gevent patching as early as possible in the application before
+    # importing more of the library internals.
+    if os.environ.get("DD_GEVENT_PATCH_ALL", "false").lower() in ("true", "1"):
+        # successfully running `gevent.monkey.patch_all()` this late into
+        # sitecustomize requires aggressive module unloading beforehand.
```

Comment:
why? can we share a little more here on why we cleanup before patching, what problem does this solve?

Reply:
this will ensure that we patch python stdlib with gevent before we load anything for ddtrace and try to create/use a thread, but how is this different from what we do today with `DD_GEVENT_PATCH_ALL=1` ?

Reply:
I feel like the order of this is wrong? don't we want to explicitly load `ddtrace` before `gevent` so we get references to the stdlib threading module and then unload it from `sys.modules` so when we import `gevent` it get's reloaded and patched, but our internal modules keep reference to the stdlib/unpatched version?

Reply:
> how is this different from what we do today with DD_GEVENT_PATCH_ALL=1 ?

It's not - it's actually exactly the current behavior. The change in this diff is that gevent patching only happens when module cloning is not happening.

> don't we want to explicitly load ddtrace before gevent

Almost. You're right about how module cloning works: import ddtrace, start threads, unload, import gevent. The one place where you're off is that the `import gevent` in this scenario happens in user code. When module cloning is active, ddtrace never imports gevent. The code block you're commenting on here is only active when module cloning *isn't*. That is, it's the current 1.x behavior gated by a new default-off feature flag.

@brettlangdon

Reply:
yeah, this is the bit I misread, I skipped the if about it when the feature is disabled.

and so we need to aggressive unload here because we need to unload `os`, `sys`, and anything they have imported before gevent loads right?

Reply:
Yes, but not `sys`. That's one of the modules that you should never unload because doing so breaks the interpreter.